oauth2_http/javatests/com/google/auth/oauth2/StsRequestHandlerTest.java

```diff
@@ -0,0 +1,311 @@
+  private static final String ACTOR_TOKEN_TYPE = "actorTokenType";
+  private static final String REQUESTED_TOKEN_TYPE = "requestedTokenType";
+  private static final String INTERNAL_OPTIONS = "internalOptions";
+  private static final String REFRESH_TOKEN = "refreshToken";
+  private static final List<String> SCOPES = Arrays.asList("scope1", "scope2", "scope3");
+
+  // Headers.
+  private static final String CONTENT_TYPE_KEY = "content-type";
+  private static final String CONTENT_TYPE = "application/x-www-form-urlencoded";
+  private static final String ACCEPT_ENCODING_KEY = "accept-encoding";
+  private static final String ACCEPT_ENCODING = "gzip";
+  private static final String CUSTOM_HEADER_KEY = "custom_header_key";
+  private static final String CUSTOM_HEADER_VALUE = "custom_header_value";
+
+  private static final String INVALID_REQUEST = "invalid_request";
+  private static final String ERROR_DESCRIPTION = "errorDescription";
+  private static final String ERROR_URI = "errorUri";
+
+  private MockStsServiceTransport transport = new MockStsServiceTransport();
+
+  @Test
+  public void exchangeToken() throws IOException {
+    StsTokenExchangeRequest stsTokenExchangeRequest =
+        StsTokenExchangeRequest.newBuilder(CREDENTIAL, SUBJECT_TOKEN_TYPE).build();
+
+    StsRequestHandler requestHandler =
+        StsRequestHandler.newBuilder(
+                TOKEN_URL, stsTokenExchangeRequest, transport.createRequestFactory())
+            .build();
+
+    StsTokenExchangeResponse response = requestHandler.exchangeToken();
+
+    // Validate response.
+    assertThat(response.getAccessToken().getTokenValue()).isEqualTo(transport.getAccessToken());
+    assertThat(response.getTokenType()).isEqualTo(transport.getTokenType());
+    assertThat(response.getIssuedTokenType()).isEqualTo(transport.getIssuedTokenType());
+    assertThat(response.getExpiresIn()).isEqualTo(transport.getExpiresIn());
+
+    // Validate request content.
+    GenericData expectedRequestContent =
+        new GenericData()
+            .set("grant_type", TOKEN_EXCHANGE_GRANT_TYPE)
+            .set("scope", CLOUD_PLATFORM_SCOPE)
+            .set("requested_token_type", DEFAULT_REQUESTED_TOKEN_TYPE)
+            .set("subject_token_type", stsTokenExchangeRequest.getSubjectTokenType())
+            .set("subject_token", stsTokenExchangeRequest.getSubjectToken());
+
+    MockLowLevelHttpRequest request = transport.getRequest();
+    Map<String, String> actualRequestContent = TestUtils.parseQuery(request.getContentAsString());
+    assertThat(actualRequestContent).isEqualTo(expectedRequestContent);
+  }
+
+  @Test
+  public void exchangeToken_withOptionalParams() throws IOException {
+    // Return optional params scope and the refresh_token.
+    List<String> scopesToReturn = new ArrayList<>();
+    scopesToReturn.add(CLOUD_PLATFORM_SCOPE);
+    scopesToReturn.addAll(SCOPES);
+
+    transport.addScopeSequence(scopesToReturn);
```

Comment:
Since `transport` seems to be re-used between tests and is mutable, will this potentially have cascading failures if tests start failing?

We can have a `@Before` method that resets the transport so it's clean between tests.

Reply:
Done.